File: ios/Classes/lib/core/models/NotificationButtonModel.swift
//
//  NotificationButtonModel.swift
//  awesome_notifications
//
//  Created by Rafael Setragni on 05/09/20.
//

import Foundation

public class NotificationButtonModel : AbstractModel {
    
    public static let TAG = "NotificationButtonModel"
    
    var key:String?
    var icon:String?
    var label:String?
    var enabled:Bool?
    var color: Int64?
    var requireInputText:Bool?
    var autoDismissible:Bool?
    var showInCompactView:Bool?
    var isDangerousOption:Bool?
    var actionType:ActionType?
    
    public init(){}
    
    public func fromMap(arguments: [String : Any?]?) -> AbstractModel? {
        if(arguments == nil){ return self }
       
        _processRetroCompatibility(fromArguments: arguments)
        
        self.key        = MapUtils<String>.getValueOrDefault(reference: Definitions.NOTIFICATION_BUTTON_KEY, arguments: arguments)
        self.icon       = MapUtils<String>.getValueOrDefault(reference: Definitions.NOTIFICATION_BUTTON_ICON, arguments: arguments)
        self.label      = MapUtils<String>.getValueOrDefault(reference: Definitions.NOTIFICATION_BUTTON_LABEL, arguments: arguments)
        self.color      = MapUtils<Int64>.getValueOrDefault(reference: Definitions.NOTIFICATION_COLOR, arguments: arguments)
        
        self.actionType = EnumUtils<ActionType>.getEnumOrDefault(reference: Definitions.NOTIFICATION_ACTION_TYPE, arguments: arguments)
        
        self.enabled    = MapUtils<Bool>.getValueOrDefault(reference: Definitions.NOTIFICATION_ENABLED, arguments: arguments)
        self.autoDismissible   = MapUtils<Bool>.getValueOrDefault(reference: Definitions.NOTIFICATION_AUTO_DISMISSIBLE, arguments: arguments)
        self.requireInputText  = MapUtils<Bool>.getValueOrDefault(reference: Definitions.NOTIFICATION_REQUIRE_INPUT_TEXT, arguments: arguments)
        self.showInCompactView = MapUtils<Bool>.getValueOrDefault(reference: Definitions.NOTIFICATION_SHOW_IN_COMPACT_VIEW, arguments: arguments)
        self.isDangerousOption = MapUtils<Bool>.getValueOrDefault(reference: Definitions.NOTIFICATION_IS_DANGEROUS_OPTION, arguments: arguments)

        return self
    }
    
    public func toMap() -> [String : Any?] {
        var mapData:[String: Any?] = [:]
        
        if(key != nil) {mapData[Definitions.NOTIFICATION_BUTTON_KEY] = self.key}
        if(icon != nil) {mapData[Definitions.NOTIFICATION_BUTTON_ICON] = self.icon}
        if(label != nil) {mapData[Definitions.NOTIFICATION_BUTTON_LABEL] = self.label}
        if(color != nil){ mapData[Definitions.NOTIFICATION_COLOR] = self.color }
        
        if(actionType != nil) {mapData[Definitions.NOTIFICATION_ACTION_TYPE] = self.actionType?.rawValue}

        if(enabled != nil) {mapData[Definitions.NOTIFICATION_ENABLED] = self.enabled}
        if(autoDismissible != nil) {mapData[Definitions.NOTIFICATION_AUTO_DISMISSIBLE] = self.autoDismissible}
        if(requireInputText != nil) {mapData[Definitions.NOTIFICATION_REQUIRE_INPUT_TEXT] = self.requireInputText}
        if(showInCompactView != nil) {mapData[Definitions.NOTIFICATION_SHOW_IN_COMPACT_VIEW] = self.showInCompactView}
        if(isDangerousOption != nil) {mapData[Definitions.NOTIFICATION_IS_DANGEROUS_OPTION] = self.isDangerousOption}
        
        return mapData
    }
    
    func _processRetroCompatibility(fromArguments arguments: [String : Any?]?){
        
        if arguments?["autoCancel"] != nil {
            Logger.w(NotificationButtonModel.TAG, "autoCancel is deprecated. Please use autoDismissible instead.")
            autoDismissible = MapUtils<Bool>.getValueOrDefault(reference: "autoCancel", arguments: arguments)
        }

        if arguments?["buttonType"] != nil {
            Logger.w(NotificationButtonModel.TAG, "buttonType is deprecated. Please use actionType instead.")            
            actionType = EnumUtils<ActionType>.getEnumOrDefault(reference: "buttonType", arguments: arguments)
        }
        
        _adaptInputFieldToRequireText()
    }
    
    func _adaptInputFieldToRequireText(){
        if actionType == ActionType.InputField {
            Logger.d(NotificationButtonModel.TAG,
                  "InputField is deprecated. Please use requireInputText instead.")
            requireInputText = true
            actionType = ActionType.SilentAction
        }
    }
    
    public func validate() throws {
        
        if(StringUtils.shared.isNullOrEmpty(key)){
            throw ExceptionFactory
                    .shared
                    .createNewAwesomeException(
                        className: NotificationButtonModel.TAG,
                        code: ExceptionCode.CODE_INVALID_ARGUMENTS,
                        message: "Button key is required",
                        detailedCode: ExceptionCode.DETAILED_INVALID_ARGUMENTS+".button.actionKey")
        }

        if(StringUtils.shared.isNullOrEmpty(label)){
            throw ExceptionFactory
                    .shared
                    .createNewAwesomeException(
                        className: NotificationButtonModel.TAG,
                        code: ExceptionCode.CODE_INVALID_ARGUMENTS,
                        message: "Button label is required",
                        detailedCode: ExceptionCode.DETAILED_INVALID_ARGUMENTS+".button.label")
        }
    }
}

File: example/android/awn_core/src/main/java/me/carda/awesome_notifications/core/models/NotificationButtonModel.java
package me.carda.awesome_notifications.core.models;

import android.content.Context;

import java.util.HashMap;
import java.util.Map;

import me.carda.awesome_notifications.core.Definitions;
import me.carda.awesome_notifications.core.enumerators.ActionType;
import me.carda.awesome_notifications.core.exceptions.AwesomeNotificationsException;
import me.carda.awesome_notifications.core.exceptions.ExceptionCode;
import me.carda.awesome_notifications.core.exceptions.ExceptionFactory;
import me.carda.awesome_notifications.core.logs.Logger;

public class NotificationButtonModel extends AbstractModel {

    private static final String TAG = "NotificationButtonModel";

    public String key;
    public String icon;
    public String label;
    public Integer color;
    public Boolean enabled;
    public Boolean requireInputText = false;
    public Boolean autoDismissible;
    public Boolean showInCompactView;
    public Boolean isDangerousOption;
    public ActionType actionType;

    @Override
    public NotificationButtonModel fromMap(Map<String, Object> arguments) {

        processRetroCompatibility(arguments);

        key               = getValueOrDefault(arguments, Definitions.NOTIFICATION_BUTTON_KEY, String.class, null);
        icon              = getValueOrDefault(arguments, Definitions.NOTIFICATION_BUTTON_ICON, String.class, null);
        label             = getValueOrDefault(arguments, Definitions.NOTIFICATION_BUTTON_LABEL, String.class, null);
        color             = getValueOrDefault(arguments, Definitions.NOTIFICATION_COLOR, Integer.class, null);
        actionType        = getValueOrDefault(arguments, Definitions.NOTIFICATION_ACTION_TYPE, ActionType.class, ActionType.Default);
        enabled           = getValueOrDefault(arguments, Definitions.NOTIFICATION_ENABLED, Boolean.class, true);
        requireInputText  = getValueOrDefault(arguments, Definitions.NOTIFICATION_REQUIRE_INPUT_TEXT, Boolean.class, false);
        isDangerousOption = getValueOrDefault(arguments, Definitions.NOTIFICATION_IS_DANGEROUS_OPTION, Boolean.class, false);
        autoDismissible   = getValueOrDefault(arguments, Definitions.NOTIFICATION_AUTO_DISMISSIBLE, Boolean.class, true);
        showInCompactView = getValueOrDefault(arguments, Definitions.NOTIFICATION_SHOW_IN_COMPACT_VIEW, Boolean.class, false);

        return this;
    }

    // Retro-compatibility with 0.6.X
    private void processRetroCompatibility(Map<String, Object> arguments){

        if (arguments.containsKey("autoCancel")) {
            Logger.w("AwesomeNotifications", "autoCancel is deprecated. Please use autoDismissible instead.");
            autoDismissible   = getValueOrDefault(arguments, "autoCancel", Boolean.class, true);
        }

        if (arguments.containsKey("buttonType")){
            Logger.w("AwesomeNotifications", "buttonType is deprecated. Please use actionType instead.");
            actionType = getValueOrDefault(arguments, "buttonType", ActionType.class, ActionType.Default);
        }

        adaptInputFieldToRequireText();
    }

    // Retro-compatibility with 0.6.X
    private void adaptInputFieldToRequireText(){
        if (actionType == ActionType.InputField) {
            Logger.d("AwesomeNotifications", "InputField is deprecated. Please use requireInputText instead.");
            actionType = ActionType.SilentAction;
            requireInputText = true;
        }
    }

    @Override
    public Map<String, Object> toMap() {
        Map<String, Object> dataMap = new HashMap<>();

        putDataOnSerializedMap(Definitions.NOTIFICATION_BUTTON_KEY, dataMap, key);
        putDataOnSerializedMap(Definitions.NOTIFICATION_BUTTON_KEY, dataMap, key);
        putDataOnSerializedMap(Definitions.NOTIFICATION_BUTTON_ICON, dataMap, icon);
        putDataOnSerializedMap(Definitions.NOTIFICATION_BUTTON_LABEL, dataMap, label);
        putDataOnSerializedMap(Definitions.NOTIFICATION_COLOR, dataMap, color);
        putDataOnSerializedMap(Definitions.NOTIFICATION_ACTION_TYPE, dataMap, actionType);
        putDataOnSerializedMap(Definitions.NOTIFICATION_ENABLED, dataMap, enabled);
        putDataOnSerializedMap(Definitions.NOTIFICATION_REQUIRE_INPUT_TEXT, dataMap, requireInputText);
        putDataOnSerializedMap(Definitions.NOTIFICATION_AUTO_DISMISSIBLE, dataMap, autoDismissible);
        putDataOnSerializedMap(Definitions.NOTIFICATION_SHOW_IN_COMPACT_VIEW, dataMap, showInCompactView);
        putDataOnSerializedMap(Definitions.NOTIFICATION_IS_DANGEROUS_OPTION, dataMap, isDangerousOption);

        return dataMap;
    }

    @Override
    public String toJson() {
        return templateToJson();
    }

    @Override
    public NotificationButtonModel fromJson(String json){
        return (NotificationButtonModel) super.templateFromJson(json);
    }

    @Override
    public void validate(Context context) throws AwesomeNotificationsException {
        if(stringUtils.isNullOrEmpty(key))
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_MISSING_ARGUMENTS,
                            "Button key is required",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".button.actionKey");

        if(stringUtils.isNullOrEmpty(label))
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_MISSING_ARGUMENTS,
                            "Button label is required",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".button.label");
    }
}
